### PR TITLE
Mayhem Integration

### PR DESCRIPTION
**Pull request by Gediminas19**

### CODE REVIEW

File: mayhem/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+## Add source code to the build stage.
+ADD . /tree-sitter
+WORKDIR /tree-sitter
+
+## build fuzzers
+RUN script/fetch-fixtures
+RUN rm -rf test/fixtures/grammars/typescript/
+RUN mkdir out && CC=clang CXX=clang++ LINK=clang++ LIB_FUZZER_PATH=/compiler-rt/lib/fuzzer/libFuzzer.a \
+    script/build-fuzzers
+
```
Comment (xansec) [inline mayhem/Dockerfile:22]: Please copy resulting binaries (and any necessary corpora) into a package stage in the Dockerfile to reduce image size.

File: mayhem/c_fuzzer.mayhemfile [hunk 2]
```diff
@@ -0,0 +1,5 @@
+project: tree-sitter
+target: c_fuzzer
+
+cmds:
```
Comment (xansec) [inline mayhem/c_fuzzer.mayhemfile:4]: Specifying dictionaries in Mayhemfiles is done with the `dictionary` field:
```
cmds:
  - cmd: <some command>
  {...}
  dictionary: /path/to/my/dictionary
```

Specifying corpora is done with the `tests` field:

```
tests:
  - file:///absolute/path/to/my/corpus
```
See https://mayhem.forallsecure.com/docs/reference/mayhemfile/#dictionary and https://mayhem.forallsecure.com/docs/reference/mayhemfile/#tests for more info.

File: script/run-fuzzer [hunk 3]
```diff
@@ -29,7 +29,7 @@ run_fuzzer() {
   # then be loaded on subsequent fuzzing runs
   mkdir -p corpus
 
-  "../../${lang}_fuzzer_${mode}" "-dict=../../${lang}.dict" "-artifact_prefix=${lang}_${mode}_" -max_len=2048 ${mode_config[$mode]} "./corpus" "$@"
```
Comment (xansec) [inline script/run-fuzzer]: Can you comment on why these changes were necessary?

**Review by xansec — CHANGES_REQUESTED** [review]

**Comment by xansec** [comment]
Please update this PR, or we will have to close.

**Comment by Gediminas19** [comment]
Sorry, I'll do it tonight

File: mayhem/c_fuzzer.mayhemfile (hunk 2 above)
Comment (Gediminas19) [reply mayhem/c_fuzzer.mayhemfile:4]: The corpus directories are actually empty (the original fuzzing in the repo had this too), so I omitted this. Let me know if I should have added them to the Mayhemfile. As for the dictionary entries, I have added them.

File: script/run-fuzzer (hunk 3 above)
Comment (Gediminas19) [reply script/run-fuzzer]: I guess they are not strictly necessary for fuzzing. Essentially I was figuring out how to build and run the fuzzers in order to write the Dockerfile and Mayhemfiles, so I was referring to the `run-fuzzer` script they provided. The script would not run correctly without this modification, which is how it made it in. I am pretty sure the script failing is a matter of it being out of date. Thus I'm not sure if my modification should be committed.

Comment (xansec) [reply script/run-fuzzer]: Looking more at this, I don't believe this change is necessary - You need to provide both a language and a "mode" (see here: https://github.com/mayhemheroes/tree-sitter/blob/80a925727bd836f11a32bb1e75196f41bffc37e2/script/run-fuzzer#L13).
Please revert this change as it doesn't seem to be blocking anything and the script doesn't appear to be used in your Dockerfile.

**Review by xansec — CHANGES_REQUESTED** [review]

File: script/run-fuzzer (hunk 3 above)
Comment (Gediminas19) [reply script/run-fuzzer]: When I ran the script, I did provide both a language and a mode. You can see that my change to the line was not deleting it, but rather removing the `_${mode}` suffix from the fuzzer executable. That is because the `build-fuzzers` script I invoke does not output fuzzer executables with that suffix. So running the original `run-fuzzer` breaks because the executable cannot be found.

But you are right, since I don't use it in the Dockerfile, I will remove it.

**Comment by xansec** [comment]
LGTM!
